sys/syslog/syslog.c

```diff
@@ -0,0 +1,621 @@
+#endif
+
+#ifndef CONFIG_SYSLOG_MAX_MSG
+#define CONFIG_SYSLOG_MAX_MSG       64
+#endif
+
+#if CONFIG_SYSLOG_HAS_OPEN
+static memarray_t mem_entries;
+static syslog_entry_t entries[CONFIG_SYSLOG_MAX_ENTRIES];
+#endif
+
+static memarray_t mem_msgs;
+static struct syslog_msg msgs[CONFIG_SYSLOG_MAX_MSG];
+
+static mutex_t lock = MUTEX_INIT;
+#endif /* CONFIG_SYSLOG_MEMARRAY */
+
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+static char _stack[SYSLOG_THREAD_STACKSIZE];
+
+#ifndef CONFIG_SYSLOG_HOSTNAME
+#define CONFIG_SYSLOG_HOSTNAME      "-"
+#endif
+
+static const char *hostname = CONFIG_SYSLOG_HOSTNAME;
+
+syslog_entry_t *syslog_default_entry;
+
+#if CONFIG_SYSLOG_HAS_OPEN
+static clist_node_t _logs;
+static mutex_t _logs_lock = MUTEX_INIT;
+#else
+static syslog_entry_t _syslog_default_entry;
+#endif
+
+#if CONFIG_SYSLOG_MEMARRAY
+static struct syslog_msg *_alloc_msg(void)
+{
+    mutex_lock(&lock);
+    struct syslog_msg *msg = memarray_alloc(&mem_msgs);
+    mutex_unlock(&lock);
+
+    return msg;
+}
+
+static void _free_msg(struct syslog_msg *msg)
+{
+    if (msg ==  NULL) {
+        return;
+    }
+    mutex_lock(&lock);
+    memarray_free(&mem_msgs, msg);
+    mutex_unlock(&lock);
+}
+
+#if CONFIG_SYSLOG_HAS_OPEN
+static syslog_entry_t *_alloc_entry(void)
+{
+    mutex_lock(&lock);
+    syslog_entry_t *entry = memarray_alloc(&mem_entries);
+    mutex_unlock(&lock);
+
+    return entry;
+}
+
+static void _free_entry(syslog_entry_t *entry)
+{
+    if (entry == NULL) {
+        return;
+    }
+    mutex_lock(&lock);
+    memarray_free(&mem_entries, entry);
+    mutex_unlock(&lock);
+}
+#endif
+#else
+static struct syslog_msg *_alloc_msg(void)
+{
+    return malloc(sizeof(struct syslog_msg));
+}
+
+static void _free_msg(struct syslog_msg *msg)
+{
+    free(msg);
+}
+
+#if CONFIG_SYSLOG_HAS_OPEN
+static syslog_entry_t *_alloc_entry(void)
+{
+    return malloc(sizeof(syslog_entry_t));
+}
+
+static void _free_entry(syslog_entry_t *entry)
+{
+    free(entry);
+}
+#endif
+#endif
+
+#if CONFIG_SYSLOG_HAS_OPEN
+syslog_entry_t *openlog(const char *ident, int option, int facility)
+{
+    syslog_entry_t *entry = _alloc_entry();
+    assert(entry);
+
+    entry->ident = ident;
+    entry->option = option;
+    entry->facility = facility;
+    entry->mask = LOG_LEVEL;
+
+    mutex_lock(&_logs_lock);
+    clist_rpush(&_logs, &entry->node);
+    mutex_unlock(&_logs_lock);
+
+    return entry;
+}
+
+void closelog(syslog_entry_t *entry)
+{
+    mutex_lock(&_logs_lock);
+    clist_remove(&_logs, &entry->node);
+    mutex_unlock(&_logs_lock);
+
+    _free_entry(entry);
+}
+#endif
+
+void syslog(syslog_entry_t *entry, int priority, const char *format, ...)
+{
+    va_list ap;
+    va_start(ap, format);
+    vsyslog(entry, priority, format, ap);
+    va_end(ap);
+}
+
+static void _set_time(struct syslog_msg *msg)
+{
+#if defined(MODULE_PERIPH_RTC)
+    rtc_get_time(&msg->time);
+    msg->time_set = true;
+#else
+    msg->time_set = false;
+#endif
```

Comment:
@maribu RTC was actually already optional ;). So I guess we can get rid of the `struct tm` in the `struct syslog_msg` and use something more lightweight.